### ISSUE TITLE
Rtree Nearest Neighbor wrong result!

### ISSUE THREAD

**Issue by carlosg-m** [Toblerity/rtree#164]
1) Set up example
```
query = gpd.GeoDataFrame(geometry=[wkt.loads('POINT (63.5812539649480755 80913.2000242236681515)')])

index = gpd.GeoDataFrame(geometry=[wkt.loads('LINESTRING (85.8859634903127471 82832.3576955195312621, 64.8852557908774514 80858.0114142517995788)'),
                                   wkt.loads('LINESTRING (85.6619588893721300 82832.6896892121876590, 39.3227658066882526 80893.9887075474689482)')])
```
2) Check data and bounds (minx, miny, maxx, maxy) 
```
print(index.geometry.bounds.iloc[0].tolist())
print(index.geometry.bounds.iloc[1].tolist())

[64.88525579087745, 80858.0114142518, 85.88596349031275, 82832.35769551953]
[39.32276580668825, 80893.98870754747, 85.66195888937213, 82832.68968921219]

print(query.geometry.bounds.iloc[0].tolist())

[63.581253964948075, 80913.20002422367, 63.581253964948075, 80913.20002422367]
```
3) Create and load Rtree
```
sindex = rtree.index.Index()
sindex.insert(0, index.geometry.bounds.iloc[0].tolist())
sindex.insert(1, index.geometry.bounds.iloc[1].tolist())
```
4) Query Rtree for nearest Neighbor (wrong index returned)
```
result = list(sindex.nearest(query.geometry.bounds.iloc[0].tolist(), 1))
print(result)

[1]
```
5) Correct result
```
result = index.geometry.map(lambda x: x.distance(query.geometry.iloc[0]))
print(result)

0     1.890925
1    23.792500
Name: geometry, dtype: float64
```

**Comment by carlosg-m** [comment]
```
from shapely.strtree import STRtree

query = gpd.GeoDataFrame(geometry=[wkt.loads('POINT (63.5812539649480755 80913.2000242236681515)')])

index = gpd.GeoDataFrame(geometry=[wkt.loads('LINESTRING (85.8859634903127471 82832.3576955195312621, 64.8852557908774514 80858.0114142517995788)'),
                                   wkt.loads('LINESTRING (85.6619588893721300 82832.6896892121876590, 39.3227658066882526 80893.9887075474689482)')])

sindex = STRtree(index.geometry.to_list())

result = sindex.nearest(query.geometry.iloc[0]).wkt

print(result)
LINESTRING (85.88596349031275 82832.35769551953, 64.88525579087745 80858.0114142518)
```

**Comment by hobu** [comment]
I'm not sure what you are expecting here. Presumably you have your bounds flipped around or something, but I don't know how geopandas works, and that's out of scope for this project anyway. Please file bugs with specific examples of Rtree-only scenarios.

**Comment by carlosg-m** [comment]
Forget about GeoPandas. The bounds are correct. 

Using shapely:
```
a = wkt.loads('LINESTRING (85.8859634903127471 82832.3576955195312621, 64.8852557908774514 80858.0114142517995788)').bounds
b = wkt.loads('LINESTRING (85.6619588893721300 82832.6896892121876590, 39.3227658066882526 80893.9887075474689482)').bounds
c = wkt.loads('POINT (63.5812539649480755 80913.2000242236681515)').bounds

print(a)
(64.88525579087745, 80858.0114142518, 85.88596349031275, 82832.35769551953)
print(b)
(39.32276580668825, 80893.98870754747, 85.66195888937213, 82832.68968921219)
print(c)
(63.581253964948075, 80913.20002422367, 63.581253964948075, 80913.20002422367)
```
```
sindex = rtree.index.Index()
sindex.insert(0, a)
sindex.insert(1, b)

print(list(sindex.nearest(c, 1)))

[1]
```

```
print(c.distance(a))
1.8909245384459328

print(c.distance(b))
23.792500147573897
```

![image](https://user-images.githubusercontent.com/55836583/90141445-0995a100-dd73-11ea-93bb-bc9a0d7aa2cb.png)

**Comment by carlosg-m** [comment]
I thought `interleaved=False` was the default behaviour? What is going on here?

![image](https://user-images.githubusercontent.com/55836583/90141964-b2dc9700-dd73-11ea-8bea-bd8b71cc613b.png)

**Comment by carlosg-m** [comment]
There is a bug here, I can guarantee it. I tested every possible combination and it doesn't work as expected.

**Comment by mwtoews** [comment]
I don't see any bug. There's only one bounding box overlap with `b` and `c` a.k.a. index 1. The bounds for `a` don't overlap with `c`.

```python
import rtree
from shapely import wkt
from shapely.geometry import box

a = wkt.loads('LINESTRING (85.8859634903127471 82832.3576955195312621, 64.8852557908774514 80858.0114142517995788)')
b = wkt.loads('LINESTRING (85.6619588893721300 82832.6896892121876590, 39.3227658066882526 80893.9887075474689482)')
c = wkt.loads('POINT (63.5812539649480755 80913.2000242236681515)')

for i, g in enumerate([a, b]):
    bbox_intersects = box(*g.bounds).intersects(c)
    print('{}: {}, {}'.format(i, bbox_intersects, g.distance(c)))

# 0: False, 1.8909245384459328
# 1: True, 23.792500147573897
```
this is expected, even though the distance to the first geometry is closest. Suggest using `nearest` to get a list of candidates that are nearest based on bounding boxes, followed by `distance` to evaluate the distance to each geometry.
```python
sindex = rtree.index.Index()
for i, g in enumerate([a, b]):
    sindex.insert(i, g.bounds)

# nearest 8 bounding boxes
nearest_list = list(sindex.nearest(c.bounds, 8))
print(nearest_list)
# [1, 0]
```
Now go thru each index to find the smallest distance to each geometry to determine the actual nearest geometry.

**Comment by carlosg-m** [comment]
@mwtoews thank you for the great explanation!

Can you explain the reason for choosing the first 8 neighbors? Is it related to the default number of leaves? Will this always guarantee the nearest neighbor?

It would be probably better to include an example like this in the documentation for users that are not acquainted with the specificities of this implementation or the Rtree structure itself.

**Comment by carlosg-m** [comment]
> I don't see any bug. There's only one bounding box overlap with `b` and `c` a.k.a. index 1. The bounds for `a` don't overlap with `c`.
> 
> ```python
> import rtree
> from shapely import wkt
> from shapely.geometry import box
> 
> a = wkt.loads('LINESTRING (85.8859634903127471 82832.3576955195312621, 64.8852557908774514 80858.0114142517995788)')
> b = wkt.loads('LINESTRING (85.6619588893721300 82832.6896892121876590, 39.3227658066882526 80893.9887075474689482)')
> c = wkt.loads('POINT (63.5812539649480755 80913.2000242236681515)')
> 
> for i, g in enumerate([a, b]):
>     bbox_intersects = box(*g.bounds).intersects(c)
>     print('{}: {}, {}'.format(i, bbox_intersects, g.distance(c)))
> 
> # 0: False, 1.8909245384459328
> # 1: True, 23.792500147573897
> ```
> 
> this is expected, even though the distance to the first geometry is closest. Suggest using `nearest` to get a list of <=8 candidates that are nearest based on bounding boxes, followed by `distance` to evaluate the distance to each geometry.
> 
> ```python
> sindex = rtree.index.Index()
> for i, g in enumerate([a, b]):
>     sindex.insert(i, g.bounds)
> 
> # nearest 8 bounding boxes
> nearest_list = list(sindex.nearest(c.bounds, num_results=8))
> print(nearest_list)
> # [1, 0]
> ```
> 
> Now go thru each index to find the smallest distance to each geometry to determine the actual nearest geometry.

The problem is that I can always create a counter example where the Rtree does not return the nearest geometry.

This is basically a heuristic or an approximation, and does not guarantee correct results, because it's providing false negatives, and excluding geometries that are in fact the nearest neighbor.
```
import rtree
from shapely import wkt
from shapely.geometry import box

lines = []
lines.append(wkt.loads('LINESTRING (85.67 82832.35, 64 80858)'))
lines.append(wkt.loads('LINESTRING (85.66 82832.36, 39 80893)'))
lines.append(wkt.loads('LINESTRING (85.65 82832.37, 38 80894)'))
lines.append(wkt.loads('LINESTRING (85.64 82832.38, 37 80895)'))
lines.append(wkt.loads('LINESTRING (85.63 82832.39, 36 80896)'))
lines.append(wkt.loads('LINESTRING (85.62 82832.40, 35 80897)'))
lines.append(wkt.loads('LINESTRING (85.61 82832.41, 34 80898)'))
lines.append(wkt.loads('LINESTRING (85.60 82832.42, 33 80899)'))
lines.append(wkt.loads('LINESTRING (84 82832.42, 36 80902)'))

point = wkt.loads('POINT (63.5812539649480755 80913.2000242236681515)')

for i, g in enumerate(lines):
    bbox_intersects = box(*g.bounds).intersects(point)
    print('{}: {}, {}'.format(i, bbox_intersects, g.distance(point)))

# 0: False, 1.0245467730315265
# 1: True, 24.08828096716559
# 2: True, 25.10168585017062
# 3: True, 26.116093479409518
# 4: True, 27.13150450944058
# 5: True, 28.147919592796637
# 6: True, 29.165339379971073
# 7: True, 30.18376451940381
# 8: True, 27.294328401853928
```
Index 0 is not included in the result even though it's the nearest neighbor (false negative):
```
sindex = rtree.index.Index()
for i, g in enumerate(lines):
    sindex.insert(i, g.bounds)

nearest_list = list(sindex.nearest(point.bounds, num_results=8))
print(nearest_list)

# [1, 2, 6, 8, 5, 3, 7, 4]
```

**This means that to obtain the first nearest neighbor and to guarantee a correct result it's necessary to compute the distance between all geometries (brute force), which makes the Rtree useless.**

**Comment by hobu** [comment]
> The problem is that I can always create a counter example where the Rtree does not return the nearest geometry.

Use a kd-tree of all of the vertices then.

> This means that to obtain the first nearest neighbor and to guarantee a correct result it's necessary to compute the distance between all geometries (brute force), which makes the Rtree useless.

"When I use a screwdriver to hammer in nails, it keeps stabbing me in the hand. This screwdriver is useless."

**Comment by carlosg-m** [comment]
> Use a kd-tree of all of the vertices then.

Using the vertices can also yield incorrect results, unless the LineString or polygon exterior is interpolated first, to guarantee a somehow consistent distance or spacing between the vertices.

For example using a k-d tree or ball tree would tell me that "Line B" is in fact the closest to "Point C", even though "Line A" is clearly closest to "Point C".

![image](https://user-images.githubusercontent.com/55836583/90319409-b9177280-df2f-11ea-89f0-bd2042aba8d8.png)

> "When I use a screwdriver to hammer in nails, it keeps stabbing me in the hand. This screwdriver is useless."

The problem here is that the screwdriver clearly says in the box that it can also hammer nails.

The first sentence in the documentation is "These features include: Nearest neighbor search".

Don't take this the wrong way. I find this lib fantastic, just not for finding nearest neighbors. Unless I'm missing something here.

**Comment by carlosg-m** [comment]
If the nearest neighbor method is applied on both directions, then the resulting index pairs are concatenated, the duplicate pairs are removed, and the distance between pairs tested, would this work?

left_geometries(right_geometries)
+
right_geometries(left_geometries)

**Comment by mwtoews** [comment]
> Can you explain the reason for choosing the first 8 neighbors? Is it related to the default number of leaves? Will this always guarantee the nearest neighbor?

8 is somewhat arbitrary. This is suitable if the geometries are neatly connected like a web of linestrings or patchwork of polygons. However, if the bounding boxes are overlapping more than 8 times in any given space, then this should be increased accordingly.

> It would be probably better to include an example like this in the documentation for users that are not acquainted with the specificities of this implementation or the Rtree structure itself.

I'll consider this!

**Comment by mwtoews** [comment]
> ```python
> nearest_list = list(sindex.nearest(point.bounds, num_results=8))
> print(nearest_list)
> 
> # [1, 2, 6, 8, 5, 3, 7, 4]
> ```

this works as expected with `num_results=9`. It would be convenient if libspatialindex could quantify the maximum number of overlaps to be used to help inform this variable.

> **This means that to obtain the first nearest neighbor and to guarantee a correct result it's necessary to compute the distance between all geometries (brute force), which makes the Rtree useless.**

It just means that you need to search for more neighbours, depending on the nature and distribution of the data. If you have millions of items, searching for the nearest 100 bounding boxes will still be drastically faster than searching the bulk lot of items.

**Comment by carlosg-m** [comment]
> this works as expected with `num_results=9` (i.e. the list includes `0`). It would be convenient if libspatialindex could quantify the maximum number of overlaps to be used to help inform this variable.

If this feature is computationally cheap and easy to implement, it would be an amazing improvement. This would basically guarantee 100% recall.

> It just means that you need to search for more neighbours, depending on the nature and distribution of the data. If you have millions of items, searching for the nearest 100 bounding boxes will still be drastically faster than searching the bulk lot of items.

Unfortunately I'm working with very noisy data, and this is actually the point, to automatically detect misplaced geometries. I've tested it and even 8 neighbors doesn't seem to be enough to get 100% recall but it's very close.

Another feature that can be really helpful is to make the rtree thread safe for reading operations. I'm using Dask to do queries in parallel and I have to make a copy of the rtree index in disk for every partition to make sure each worker/core only has access to it, and avoid multiple reads.
